site/app/libraries/MainRouter.php

```diff
@@ -0,0 +1,141 @@
+                $this->request = Request::create(
+                    $this->core->buildNewUrl(['navigation', 'no_access']),
+                    'GET'
+                );
+                $this->parameters = $this->matcher->matchRequest($this->request);
+            }
+        }
+        elseif ($this->core->getConfig()->isCourseLoaded()
+            && !$this->core->getAccess()->canI("course.view", ["semester" => $this->core->getConfig()->getSemester(), "course" => $this->core->getConfig()->getCourse()])
+            && !$this->endsWith($this->parameters['_controller'], 'AuthenticationController')) {
+            $this->request = Request::create(
+                $this->core->buildNewUrl(['navigation', 'no_access']),
+                'GET'
+            );
+            $this->parameters = $this->matcher->matchRequest($this->request);
+        }
+
+        // TODO: log
+
+        if(!$this->core->getConfig()->isCourseLoaded()) {
+            if ($this->logged_in){
+                if (isset($this->parameters['_method']) && $this->parameters['_method'] === 'logout'){
+                    $this->request = Request::create(
+                        $this->core->buildNewUrl(['authentication', 'logout']),
+                        'GET'
+                    );
+                    $this->parameters = $this->matcher->matchRequest($this->request);
+                }
+                else {
+                    $this->request = Request::create(
+                        $this->core->buildNewUrl(['home']),
+                        'GET'
+                    );
+                    $this->parameters = $this->matcher->matchRequest($this->request);
+                }
+            }
+            else {
+                $this->request = Request::create(
+                    $this->core->buildNewUrl(['authentication']),
+                    'GET'
+                );
+                $this->parameters = $this->matcher->matchRequest($this->request);
+            }
+        }
+
+        if (empty($this->parameters['_controller']) && $this->core->getUser() !== null) {
+            if ($this->core->getConfig()->isCourseLoaded()) {
+                $this->request = Request::create(
+                    $this->core->buildNewUrl(['navigation']),
+                    'GET'
+                );
+                $this->parameters = $this->matcher->matchRequest($this->request);
+            }
+            else {
+                $this->request = Request::create(
+                    $this->core->buildNewUrl(['home']),
+                    'GET'
+                );
+                $this->parameters = $this->matcher->matchRequest($this->request);
+            }
+        }
+    }
+
+    private function endsWith($string, $ending) {
+        $str_len = strlen($string);
+        $ending_len = strlen($ending);
+        if ($ending_len > $str_len) return false;
+        return substr_compare($string, $ending, $str_len - $ending_len, $ending_len) === 0;
+    }
```

Comment:
Can you not just use [Utils::endsWith](https://github.com/Submitty/Submitty/blob/master/site/app/libraries/Utils.php#L127-L129) for this?

Reply:
Sorry, I did not know there is a `Utils::endsWith`. I have made the changes.